modules/module-postgres-storage/src/storage/PostgresBucketStorageFactory.ts

```diff
@@ -0,0 +1,499 @@
+import * as framework from '@powersync/lib-services-framework';
+import { storage, sync, utils } from '@powersync/service-core';
+import * as pg_wire from '@powersync/service-jpgwire';
+import * as sync_rules from '@powersync/service-sync-rules';
+import crypto from 'crypto';
+import { wrapWithAbort } from 'ix/asynciterable/operators/withabort.js';
+import { LRUCache } from 'lru-cache/min';
+import * as timers from 'timers/promises';
+import * as uuid from 'uuid';
+
+import * as lib_postgres from '@powersync/lib-service-postgres';
+import { models, NormalizedPostgresStorageConfig } from '../types/types.js';
+
+import { NOTIFICATION_CHANNEL, STORAGE_SCHEMA_NAME } from '../utils/db.js';
+import { notifySyncRulesUpdate } from './batch/PostgresBucketBatch.js';
+import { PostgresSyncRulesStorage } from './PostgresSyncRulesStorage.js';
+import { PostgresPersistedSyncRulesContent } from './sync-rules/PostgresPersistedSyncRulesContent.js';
+
+export type PostgresBucketStorageOptions = {
+  config: NormalizedPostgresStorageConfig;
+  slot_name_prefix: string;
+};
+
+export class PostgresBucketStorageFactory
+  extends framework.DisposableObserver<storage.BucketStorageFactoryListener>
+  implements storage.BucketStorageFactory
+{
+  readonly db: lib_postgres.DatabaseClient;
+  public readonly slot_name_prefix: string;
+
+  protected notificationConnection: pg_wire.PgConnection | null;
+  private sharedIterator = new sync.BroadcastIterable((signal) => this.watchActiveCheckpoint(signal));
+
+  private readonly storageCache = new LRUCache<number, storage.SyncRulesBucketStorage>({
+    max: 3,
+    fetchMethod: async (id) => {
+      const syncRulesRow = await this.db.sql`
+        SELECT
+          *
+        FROM
+          sync_rules
+        WHERE
+          id = ${{ value: id, type: 'int4' }}
+      `
+        .decoded(models.SyncRules)
+        .first();
+      if (syncRulesRow == null) {
+        // Deleted in the meantime?
+        return undefined;
+      }
+      const rules = new PostgresPersistedSyncRulesContent(this.db, syncRulesRow);
+      return this.getInstance(rules);
+    },
+    dispose: (storage) => {
+      storage[Symbol.dispose]();
+    }
+  });
+
+  constructor(protected options: PostgresBucketStorageOptions) {
+    super();
+    this.db = new lib_postgres.DatabaseClient({
+      config: options.config,
+      schema: STORAGE_SCHEMA_NAME,
+      notificationChannels: [NOTIFICATION_CHANNEL]
+    });
+    this.slot_name_prefix = options.slot_name_prefix;
+
+    this.db.registerListener({
+      connectionCreated: async (connection) => this.prepareStatements(connection)
+    });
+    this.notificationConnection = null;
+  }
+
+  async [Symbol.dispose]() {
```

Comment:
This should use `Symbol.asyncDispose` - `Symbol.dispose` is for synchronous disposing.

Reply:
That is better. I've updated the `BucketStorageFactory` interface to use `AsyncDisposable` and all the relevant usages to use `await using`